src/core/net/dnssd_server.cpp

```diff
@@ -1103,8 +1108,32 @@ void Server::FinalizeQuery(QueryTransaction &aQuery, Header::Response aResponseC
     OT_ASSERT(sdType != kDnsQueryNone);
     OT_UNUSED_VARIABLE(sdType);
 
-    mQueryUnsubscribe(mQueryCallbackContext, name);
     aQuery.Finalize(aResponseCode, mSocket);
+    // We are the last subscribing this name
+    if (!HasSameQuery(name))
+    {
+        mQueryUnsubscribe(mQueryCallbackContext, name);
+    }
+}
+
+bool Server::HasSameQuery(const char *aName)
```

Comment:
We should follow the name searching rules implemented in `GetQueryType`?

Currently `GetQueryType` searches for PTR/TXT/SRV questions first, and then AAAA/A questions.
Otherwise the logic is not strictly correct.

Thoughts?

Reply:
Updated code to check the query type as well.